libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Extensions/RestClient.swift

```
@@ -136,6 +159,70 @@ extension RestClient {
         })
     }
   
+    // MARK: Record Convience API - Pure Swift 4+
+    
+    /// This method provides a reusuable, generic pipeline for retrieving records
+    ///   from Salesforce. It relys on Swift Generics, and type inference to determine what
+    ///  models to create.
+    ///
+    /// Given a model object - Contact, you can use this method like this:
+    ///   RestClient.shared.fetchRecords(ofModelType: ModelName.self, forRequest: request) { result in
+    ///       switch result {
+    ///       case .success(let records):
+    ///           do something with your array of model objects
+    ///       case .failure(let error):
+    ///           print(error)
+    ///       }
+    ///     }
+    ///
+    /// This method relies on the passed parameter ofModelType to infer the generic Record's
+    /// concrete type.
+    func fetchRecords<Record: Decodable>(ofModelType: Record.Type, forRequest request: RestRequest,
+                                       _ completionBlock: @escaping (Result<[Record], RestClientError>) -> Void) {
+      guard request.isQueryRequest else { return }
+      RestClient.shared.send(request: request) { result in
+          switch result {
+              case .success(let response):
+                do {
+                  let decoder = JSONDecoder()
+                  let wrapper = try decoder.decode(QueryResponseWrapper<Record>.self, from: response.asData())
+                  completionBlock(.success(wrapper.records))
+                } catch {
+                  completionBlock(.success([Record]()))
```

@codefriar I think we should either rethrow the error or not catch the decoder issues.

---

Since the underlying Send Method doesn't throw, I can't throw here. I'm now returning an empty wrapper object.